Initial implementation of EnaSolar integration

geustace opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Initial implementation of an integration to add the EnaSolar Solar Inverter.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
https://github.com/home-assistant/home-assistant.io/pull/19212
https://github.com/home-assistant/brands/pull/2759

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

geustace committed:
Initial implementation of EnaSolar integration

geustace committed:
Disable not-callable

SteveEasley commented on homeassistant/components/enasolar/sensor.py:
```diff
@@ -0,0 +1,248 @@
+    if hass.is_running:
+        start_update_interval()
+    else:
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_update_interval)
```
You might want to look into using DataUpdateCoordinator for official polling support (https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities). I haven't used it myself, but looks like it has rate limiting features as well.

geustace replied:
I was hoping to leave this to the next version, I have been trying to get this integration over the line now for many months and what is in place currently works. And yes I know the dev team a REALLY busy.

SteveEasley commented on homeassistant/components/enasolar/sensor.py:
```diff
@@ -0,0 +1,248 @@
+
+    _LOGGER.debug("   Polling between %s and %s", enasolar.sun_up, enasolar.sun_down)
+    _LOGGER.debug(
+        "   Max Output: %s, DC Strings: %s, Capability: %s",
```
Is there a reason for prefixing debug messages with whitespace?

geustace replied:
Just personal preference, I wanted the block of debug messages to standout while I was debugging

gjohansson-ST replied:
CI will complain so get it fixed

geustace replied:
Have removed

SteveEasley reviewed (commented):
Code quality looks great. Just a few comments.

geustace commented on homeassistant/components/enasolar/__init__.py:
```diff
@@ -0,0 +1,45 @@
+    except Exception as conerr:
+        raise ConfigEntryNotReady(
+            f"Connection to EnaSolar Inverter '{host}' failed ({conerr})"
+        ) from conerr
```
In the case of an inverter, it is usually available 24x7, a failure would likely be the wifi in my house was down for a while.

SteveEasley replied:
Yep which is exactly what ConfigEntryNotReady is for. Just need to make sure you are only catching exception types that might go away on their own (timeouts etc).

gjohansson-ST commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+            if serial == entry.unique_id
+        )
+
+    async def _try_connect(self, host):
```
```suggestion
    async def _try_connect(self, host) -> str | None:
```

geustace replied:
Noted, updated

gjohansson-ST commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
```
```suggestion
    def async_get_options_flow(config_entry) -> FlowHandler:
```

gjohansson-ST replied:
Seems missing in many places

geustace replied:
Am new to python but will see if I can clean up.

gjohansson-ST commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+                        return self.async_abort(reason="already_configured")
+                    await self.async_set_unique_id(self.get_serial_no())
+                    return await self.async_step_inverter()
+        else:
+            user_input = {}
+            user_input[CONF_NAME] = DEFAULT_NAME
+            user_input[CONF_HOST] = DEFAULT_HOST
```
Handle this is schema instead

geustace replied:
Agree, was duplicated.  But need to initialise from saved options so rewrote.

gjohansson-ST commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+        # Only exposes methods, no significant
+        # code is executed during instantiation
+        self._enasolar = pyenasolar.EnaSolar()
+        self._data: dict[str, Any] = {}
```
Why is this needed when we have user_input already?

geustace replied:
Not sure what you mean, some default data is extracted from the inverter so need to be able to get them.

gjohansson-ST commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+            last_step=False,
+        )
+
+    async def async_step_inverter(self, user_input=None):
+        """Give the user the opportunities to override inverter config."""
```
Should this not move to the option flow instead? Needed to setup the integration?

geustace replied:
No, the basic configuration of the inverter is fixed, if the hardware is modified, the integration should be deleted and re-added. Only the polling window is able to be updated once the inverter is added.

gjohansson-ST commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+    def __init__(self, config_entry):
+        """Initialize EnaSolar options flow."""
+        self.config_entry = config_entry
+        self._options = dict(config_entry.options)
```
```suggestion
        self._options = config_entry.options
```
Should already be a dict?

geustace replied:
Yep, updated

geustace replied:
If I remove this, the the update method isn't available for _options.

gjohansson-ST commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+            if input_valid:
+                self._options.update(user_input)
+                return self.async_create_entry(title="", data=self._options)
+        else:
+            user_input = {}
+            if not self._options:
+                user_input[CONF_SUN_UP] = DEFAULT_SUN_UP
+                user_input[CONF_SUN_DOWN] = DEFAULT_SUN_DOWN
```
Suggest to handle it in the schema instead with user_input.get(CONF_SUN_UP, DEFAULT_SUN_UP)

geustace replied:
Rewritten but still not sure am doing it right

gjohansson-ST replied:
you can remove this else completely and change as per below

gjohansson-ST commented on homeassistant/components/enasolar/sensor.py:
```diff
@@ -0,0 +1,248 @@
+            else:
+                hass_data_sensors.append(
+                    EnaSolarSensor(
+                        sensor, config_entry.data[CONF_NAME], enasolar.serial_no
```
This seems identical to the one above?

geustace replied:
Nope, the data and meter values have different polling intervals.

gjohansson-ST commented on homeassistant/components/enasolar/sensor.py:
```diff
@@ -0,0 +1,248 @@
+        self._attr_should_poll = False
+
+    @property
+    def name(self):
```
Use _attr_name * instead.
Goes also for the others below

geustace replied:
Sorry, don't understand this comment.

SteveEasley replied:
I think thats the preferred method of setting attributes these days.

https://github.com/home-assistant/core/blob/caaa1b32c1808e0fd23f9ab42e1f0b481110883d/homeassistant/helpers/entity.py#L301-L308. https://github.com/home-assistant/core/blob/caaa1b32c1808e0fd23f9ab42e1f0b481110883d/homeassistant/components/sensor/__init__.py#L251-L263.

gjohansson-ST commented on homeassistant/components/enasolar/strings.json:
```diff
@@ -0,0 +1,45 @@
+          "sun_up": "The time that polling will start each day",
+          "sun_down": "The time that polling will cease each day"
+        },
+        "title": "EnaSolar Polling"
```
Can be removed. It takes the integration name

geustace replied:
I don't want it taking the integration name as the options available are only the polling window.

gjohansson-ST commented on homeassistant/components/enasolar/translations/en.json:
```diff
@@ -0,0 +1,45 @@
+        "abort": {
+            "already_configured": "Device is already configured"
+        },
+        "error": {
```
Isn't `unknown` missing

geustace replied:
Agreed, added

gjohansson-ST reviewed (changes requested):
Some comments

geustace commented on homeassistant/components/enasolar/__init__.py:
```diff
@@ -0,0 +1,45 @@
+
+from .const import DOMAIN, ENASOLAR
+
+PLATFORMS = ["sensor"]
```
Updated

gjohansson-ST replied:
You have not pushed your changes so nothing to be looked at

geustace commented on homeassistant/components/enasolar/__init__.py:
```diff
@@ -0,0 +1,45 @@
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Set up the config entry."""
+    host = config_entry.data[CONF_HOST]
+    _LOGGER.debug("Instantiate an EnaSolar Inverter at '%s'", host)
```
Was helpful when debugging, I am new to HA and Python, but have removed

geustace commented on homeassistant/components/enasolar/__init__.py:
```diff
@@ -0,0 +1,45 @@
+            f"Connection to EnaSolar Inverter '{host}' failed ({conerr})"
+        ) from conerr
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][config_entry.entry_id] = {ENASOLAR: enasolar}
```
Agree more concise, updated

gjohansson-ST commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+                _errors[CONF_HOST] = "invalid_host"
+            else:
+                self._data[CONF_HOST] = user_input[CONF_HOST]
+                self._data[CONF_NAME] = user_input[CONF_NAME]
```
Remove the name and use the host as name

geustace replied:
Host is converted to IP address.  I don't want the inverter named as an IP would rather have text.

gjohansson-ST replied:
I can understand but essentially users should be asked as little as possible. There is no name or something good that comes back from the api that can be used as name instead?

geustace replied:
No there isn't, and as it is possible to have multiple inverters that needed even be at the same location, allowing the user to set a name seemed sensible to me. There isn't an API either, the data is either pulled from some JScript on the webpage or directly accessing the pages that have the data.

geustace replied:
It is an optional field, so the user can just ignore if it makes more sense and they only have the one inverter

gjohansson-ST replied:
I get it but the config entry could be with the IP address and the user can always rename the device as they wish.
Also `DEFAULT_NAME` is missing in the config schema so your code would fail if someone would not type a name.

geustace replied:
The DEFAULT_NAME is an empty string, I don't name my inverter and the code seems to work just fine.

gjohansson-ST replied:
The name should be the host and the user should not be asked to provide a name. If the user wants to have a specific name instead of an IP address then the user can rename the entry in the UI after the integration has been setup.

geustace replied:
We might have to agree to differ on this one.  Renaming the device does not rename the sensors it is parent of and as there are 13 or more of them renaming each individually is a hassle. Currently, if no name is given, the sensors use the prefix 'enasolar', if a name is provided when the device is added, the sensors are named with that name as a prefix. I have just removed the additional 'enasolar' prefix in this case as it made the sensor name a bit unwieldy.

gjohansson-ST commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,218 @@
+            step_id="init",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_SUN_UP, default=user_input[CONF_SUN_UP]
+                    ): cv.string,
```
```suggestion
                    vol.Required(
                        CONF_SUN_UP, default=user_input.get(CONF_SUN_UP, DEFAULT_SUN_UP)
                    ): cv.string,
```

geustace replied:
OK, I think I have this right now.

gjohansson-ST reviewed (changes requested):
Additional comments waiting for push

geustace committed:
Most of the suggested changes

geustace committed:
Add missing parameter from start_update interval

geustace committed:
Missed parameter

geustace committed:
another missed parameter

geustace committed:
Use _attr_name

geustace committed:
Remove leading whitespace

geustace committed:
simplify the sensor name prefix

ezlo-picori commented on homeassistant/components/enasolar/const.py:
```diff
@@ -0,0 +1,53 @@
+SCAN_DATA_MIN_INTERVAL = 600
+SCAN_MAX_INTERVAL = 3600
+
+MAX_OUTPUT = [1.5, 2.0, 3.0, 3.8, 4.0, 5.0]
+DC_STRINGS = [1, 2]
```
This integration has clearly been iterated over so just a minor comment: I personnaly feel more comfortable using immutables for constant values. So `tuple` rather than `list` would make sense here I believe.

geustace committed:
change constant from list to tuple

frenck commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+        self._enasolar = pyenasolar.EnaSolar()
+        self._data: dict[str, Any] = {}
+
+    def _conf_for_inverter_exists(self, serial) -> bool:
```
partially typed, please type methods at full

geustace replied:
Ok, actioned

frenck commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+            if serial == entry.unique_id
+        )
+
+    async def _try_connect(self, host) -> str | None:
```
partially typed, please type methods at full

geustace replied:
OK, actioned

frenck commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry) -> OptionsFlow:
```
partially typed, please type methods at full

geustace replied:
Ok, actioned

frenck commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+        """Get options flow for this handler."""
+        return EnaSolarOptionsFlowHandler(config_entry)
+
+    async def async_step_user(self, user_input=None) -> FlowResult:
```
partially typed, please type methods at full

geustace replied:
Ok, actioned

frenck commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+            last_step=False,
+        )
+
+    async def async_step_inverter(self, user_input=None) -> FlowResult:
```
partially typed, please type methods at full

geustace replied:
Ok, actioned

frenck commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+class EnaSolarOptionsFlowHandler(config_entries.OptionsFlow):
+    """Allow Polling window to be updated."""
+
+    def __init__(self, config_entry):
```
lets add typing here

geustace replied:
Ok, actioned

frenck commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+        self.config_entry = config_entry
+        self._options = dict(config_entry.options)
+
+    async def async_step_init(self, user_input=None) -> FlowResult:
```
partially typed, please type methods at full

geustace replied:
Ok, actioned

frenck commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+                    vol.Required(
+                        CONF_DC_STRINGS, default=user_input[CONF_DC_STRINGS]
+                    ): vol.In(DC_STRINGS),
+                    vol.Required(
+                        CONF_CAPABILITY, default=user_input[CONF_CAPABILITY]
+                    ): vol.Coerce(int),
```
Honestly, we cannot ask an end-user for entering a bitmask in the UI.
Ever.

geustace replied:
Ok, would an alternative be a collection of check boxes and I build up the mask ? It is unlikely that a user would want to change this but it does determine which meters are going to be present and hence need to have sensors created.

gjohansson-ST replied:
A multi-select would seem a good choice perhaps.

geustace replied:
Can you point me at an example of how to do a multi-select, I cant seem to find anything.

gjohansson-ST replied:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/demo/config_flow.py#L94

frenck commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+            step_id="inverter",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_MAX_OUTPUT, default=user_input[CONF_MAX_OUTPUT]
+                    ): vol.In(MAX_OUTPUT),
+                    vol.Required(
+                        CONF_DC_STRINGS, default=user_input[CONF_DC_STRINGS]
+                    ): vol.In(DC_STRINGS),
```
Are these required for setup? Or options? Are they even really needed at all?

geustace replied:
The determine which sensors need to be created.  I added them as I anted the ability to override what I managed to guess from interpreting some obfuscated JScript that was returned from the inverted. If I had it wrong, I wanted to still be able to create the sensors.

frenck commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+        _errors = {}
+        if user_input is not None:
+            input_valid = True
+            if dt_util.parse_time(user_input[CONF_SUN_UP]) is None:
+                _errors[CONF_SUN_UP] = "time_invalid"
+                input_valid = False
+            if dt_util.parse_time(user_input[CONF_SUN_DOWN]) is None:
+                _errors[CONF_SUN_DOWN] = "time_invalid"
+                input_valid = False
+            if input_valid:
+                if dt_util.parse_time(user_input[CONF_SUN_UP]) >= dt_util.parse_time(user_input[CONF_SUN_DOWN]):  # type: ignore
+                    _errors[CONF_SUN_DOWN] = "time_range"
+                    input_valid = False
```
This seems to be used to adjust polling. Why?

geustace replied:
The inverter's web server crashes if polled to frequently for too long from the same IP address.  My solution was to throttle back the polling frequency of both meters and cumulative data counters separately and to not poll when it was  dark (i.e. no sun, no generation). This is an Enasolar thing not an HA issue.

geustace replied:
I did consider using the sun integration but for a first try just making them static but configurable seemed simplest .

frenck commented on homeassistant/components/enasolar/sensor.py:
```diff
@@ -0,0 +1,250 @@
+
+    enasolar = hass.data[DOMAIN][config_entry.entry_id][ENASOLAR]
+
+    if config_entry.options != {}:
+        enasolar.sun_up = dt_util.parse_time(config_entry.options[CONF_SUN_UP])
+        enasolar.sun_down = dt_util.parse_time(config_entry.options[CONF_SUN_DOWN])
+    else:
+        enasolar.sun_up = dt_util.parse_time(DEFAULT_SUN_UP)
+        enasolar.sun_down = dt_util.parse_time(DEFAULT_SUN_DOWN)
```
```suggestion
    enasolar.sun_up = dt_util.parse_time(config_entry.options.get(CONF_SUN_UP, DEFAULT_SUN_UP)
    enasolar.sun_down = dt_util.parse_time(config_entry.options.get(CONF_SUN_DOWN, DEFAULT_SUN_DOWN)
```

geustace replied:
Ok, just showing how much a python newbie I am. Actioned

frenck commented on homeassistant/components/enasolar/sensor.py:
```diff
@@ -0,0 +1,250 @@
+            )
+        return values
+
+    def start_update_interval(event):  # pylint: disable=unused-argument
+        """Start the update interval scheduling."""
+        config_entry.async_on_unload(
+            async_track_time_interval_backoff(
+                hass, async_enasolar_meters, SCAN_METERS_MIN_INTERVAL
+            )
+        )
+        config_entry.async_on_unload(
+            async_track_time_interval_backoff(
+                hass, async_enasolar_data, SCAN_DATA_MIN_INTERVAL
+            )
+        )
```
Please implement a `DataUpdateCoordinator` instead.

geustace replied:
I was hoping to leave that to the next version as I am still struggling with all the rest of this implementation and haven't understood how DataUpdateCoordinator would work.

frenck commented on homeassistant/components/enasolar/sensor.py:
```diff
@@ -0,0 +1,250 @@
+        self._attr_native_unit_of_measurement = ENASOLAR_UNIT_MAPPINGS[self.sensor.unit]
+        self._attr_should_poll = False
+
+    @property
+    def name(self) -> str | None:
+        """Return the name of the sensor."""
+        if hasattr(self, "_attr_name"):
+            return self._attr_name
+        return None
+
```
```suggestion
```

geustace replied:
So the property name is not needed?  I have actioned.

frenck commented on homeassistant/components/enasolar/sensor.py:
```diff
@@ -0,0 +1,250 @@
+        """Return the current sensor value."""
+        return self._native_value
+
+    @property
+    def device_class(self):
+        """Return the device class the sensor belongs to."""
+        if self._attr_native_unit_of_measurement == POWER_KILO_WATT:
+            return DEVICE_CLASS_POWER
+        if self._attr_native_unit_of_measurement == ENERGY_KILO_WATT_HOUR:
+            return DEVICE_CLASS_ENERGY
+        if self._attr_unit_of_measurement in (TEMP_CELSIUS, TEMP_FAHRENHEIT):
+            return DEVICE_CLASS_TEMPERATURE
+        return None
```
Things like these makes me wonder if you should implement `SensorEntityDescription`s instead?

geustace replied:
I started this integration by basing it on an exisiting one 'saj' as it seemed to do the sorts of things needed.  It obviously was not a good choice as it doesn't conform with the coding guidelines that now are required. I freely admit to not understanding the internals of HA and how the various things interact, my goal was to get something up and running I could use and then I decided that it would possibly be useful for other people as well as EnaSolar was a popular inverter brand in the Southern Hemisphere.

frenck commented on homeassistant/components/enasolar/sensor.py:
```diff
@@ -0,0 +1,250 @@
+    @property
+    def unique_id(self):
+        """Return a unique identifier for this sensor."""
+        return f"{self.serial_no}_{self.sensor.name}"
```
What is name?

geustace replied:
I assumed that this was the unique id of the sensor, not the device so in order to create something unique I used the devices serial number and the sensor's name e.g. energy_today_total => 1234567890_energy_today_total.

frenck reviewed (changes requested):
Some initial comments; there are more, but since some bigger suggestions are made in here I think some of the comments I didn't make are gone anyways.

geustace commented on homeassistant/components/enasolar/__init__.py:
```diff
@@ -0,0 +1,43 @@
+            f"Connection to EnaSolar Inverter '{host}' failed ({conerr})"
+        ) from conerr
+
+    hass.data.setdefault(DOMAIN, {})[config_entry.entry_id] = {ENASOLAR: enasolar}
```
Actioned.

geustace commented on homeassistant/components/enasolar/__init__.py:
```diff
@@ -0,0 +1,43 @@
+
+async def async_unload_entry(hass, config_entry):
+    """Unload a config entry."""
+    return await hass.config_entries.async_unload_platforms(config_entry, PLATFORMS)
```
I think entry in 38 and 39 should be config_entry, so have actioned with that change.

geustace replied:
I think entry in 38 and 39 should be config_entry, so have actioned with that change.

geustace commented on homeassistant/components/enasolar/config_flow.py:
```diff
@@ -0,0 +1,216 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+def _get_ip(host):
```
Actioned

geustace commented on homeassistant/components/enasolar/strings.json:
```diff
@@ -0,0 +1,46 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "Setup EnaSolar Inverter connection",
```
OK, actioned

geustace commented on homeassistant/components/enasolar/strings.json:
```diff
@@ -0,0 +1,46 @@
+        "data": {
+          "host": "FQDN or IP Address of the Inverter",
+          "name": "A name for the Inverter (optional)"
+        },
+        "description": "Enter the connection data"
```
OK, actioned

geustace commented on homeassistant/components/enasolar/__init__.py:
```diff
@@ -0,0 +1,43 @@
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
```
Ok, actioned

geustace committed:
Modifications recommended by frenk

geustace committed:
convert capability to a multi-select

geustace commented:
I have changed the user input for capability and updated the tests, but the test fails and I don't know what to do to fix.  Any suggestions?
        # Use the capability bits from the Inverter. This assumes it
        # had been possible to actually scrape them from the jScript
        capabilities = []
        caps = self.get_capability()
>       if caps & HAS_POWER_METER:
E       TypeError: unsupported operand type(s) for &: 'NoneType' and 'int'

homeassistant/components/enasolar/config_flow.py:167: TypeError

ezlo-picori commented:
> I have changed the user input for capability and updated the tests, but the test fails and I don't know what to do to fix. Any suggestions? # Use the capability bits from the Inverter. This assumes it # had been possible to actually scrape them from the jScript capabilities = [] caps = self.get_capability()
> 
> > ```
> >   if caps & HAS_POWER_METER:
> > ```
> 
> E TypeError: unsupported operand type(s) for &: 'NoneType' and 'int'
> 
> homeassistant/components/enasolar/config_flow.py:167: TypeError

You forgot the patch context at the beginning of `test_inverter`:

```python
async def test_inverter(hass):
    """Test user config."""
    flow = init_config_flow(hass)

    #   Firstly with no data should return the form
    with patch(
        "pyenasolar.EnaSolar.get_capability",
        return_value=CAPABILITY,
    ):
        result = await flow.async_step_inverter(None)
```

The call to `flow.async_step_inverter(None)` performs a call to `self.get_capability()` which, without mocking, returns `None`.
I advise you to fully type your `pyenasolar` package. For instance, the attribute `EnaSolar.capability` should have for type `int | None` but `get_capability(self)` should only return a `int` and raise an error if the attribute is still `None`. This way you would never retrieve a non-expected type and if a method is called unmocked in tests, you will get an exception right at the source of the issue.

BTW, the following methods are not required: you can directly patch the corresponding methods in `pyenasolar.EnaSolar` class as I have done just before.

```python
    def get_serial_no(self) -> str:
        """Needed to mock serial_no when running tests."""
        return self._enasolar.get_serial_no()

    def get_capability(self) -> int:
        """Needed to mock capability when running tests."""
        return self._enasolar.get_capability()
```

I now get the following error when running your tests:

```
        if user_input is not None:
            # CAPABILITY is a 9 bit value, bits 1-3 and 9 representing
            # what features the Innverter has.
            cap: int = 0
>           if "power" in user_input[CONF_CAPABILITY]:
E           TypeError: argument of type 'int' is not iterable

homeassistant/components/enasolar/config_flow.py:149: TypeError
```
With `user_input[CONF_CAPABILITY] = 256`, testing `"power" in 256` does not make sense but I guess this is work in progress and you were concentrating on the error you mentionned.

geustace committed:
fix up test, type pyenasolar

frenck reviewed (changes requested):
The upstream library uses webscraping:

https://github.com/geustace/pyenasolar/blob/314ef1bc674cdf62dc43adb1076bcc383e350baf/pyenasolar/__init__.py#L176-L184

This is not allowed for integrations in Home Assistant.
See ADR-0004 for details: <https://github.com/home-assistant/architecture/blob/master/adr/0004-webscraping.md>

geustace commented:
I have been informed by @frenck that this PR will never be accept due to the pyenasolar library using web scraping to extract the initial config from the inverter. This is disappointing as there is no other way to get the information.  This model of inverter is no longer manufactured or supported by the vendor so there is no chance that the inverter's internal web server code will ever be updated. The meters and other data comes from XML pages served by the inverter.

If I don't hear from anyone that there is maybe a way forward, I'll close this request and try to make my code available as a custom integration.

gjohansson-ST commented:
Given that it's also not supported or maintained from the vendor maybe another reason not to move it in here?
Publish it as a custom integration instead, nothing stopping that.

geustace commented:
Will make it a custom Integration instead.
BTW: I'd like to thank all those who have commented during the development, my python knowledge has improved and my HASS understanding is much better than it was.
